lib/widgets/medication_details.dart

```diff
@@ -0,0 +1,670 @@
+            },
+            items: (items.where((element) => durationUnits.contains(element['name'])))
+                .map<DropdownMenuItem<String>>((vt) =>
+                    DropdownMenuItem<String>(
+                        value: vt['name'],
+                        child: Text(vt['name'] ?? 'unknown')))
+                .toList(),
+            validator: (value) => value == null ? durationUnitRequired : null,
+          );
+        },
+        valueListenable: _typeChangeAllowed,
+      ));
+  }
+
+  Padding _dosingInstructionsRow(String lblTitle, List<dynamic> items) {
+    return Padding(
+        padding: const EdgeInsets.symmetric(horizontal: 16.0),
+        child:
+            Row(mainAxisAlignment: MainAxisAlignment.spaceBetween, children: [
+          Text(lblTitle),
+          Container(
+              alignment: Alignment.centerRight,
+              width: 200,
+              child: ValueListenableBuilder<bool>(
+                builder:
+                    (BuildContext context, bool allowChange, Widget? child) {
+                  return DropdownButtonFormField<String>(
+                    value: _selectedDosingInstructions,
+                    alignment: Alignment.centerRight,
+                    onChanged: (newvalue) {
+                      setState(() {
+                        _selectedDosingInstructions = newvalue!;
+                      });
+                    },
+                    items: (items)
+                        .map<DropdownMenuItem<String>>((vt) =>
+                            DropdownMenuItem<String>(
+                                value: vt['name'],
+                                child: Text(vt['name'] ?? 'unknown')))
+                        .toList(),
+                    validator: (value) => value == null ? dosingInstructionRequired : null,
+                  );
+                },
+                valueListenable: _typeChangeAllowed,
+              ))
+        ]));
+  }
+
+  Padding _numericRow(String lblTitle, TextEditingController controller) {
+    return Padding(
+        padding: const EdgeInsets.symmetric(horizontal: 16.0),
+        child:
+            Row(mainAxisAlignment: MainAxisAlignment.spaceBetween, children: [
+          Text(lblTitle),
+          Row(
+            children: [
+              Container(
+                padding: EdgeInsets.only(right: 10),
+                alignment: Alignment.centerRight,
+                width: 50,
+                child: TextFormField(
+                    keyboardType: TextInputType.number,
+                    controller: controller,
+                    validator: (value) => value!.isEmpty ? controller == _doseController ? doseRequired : durationRequired : null,
+                    onTapOutside: (_) {
+                      setState(() {
+                        if (controller == _doseController) {
+                          _medication.dosingInstructions?.dose =
+                              double.tryParse(_doseController.text);
```

Comment:
none of the double or int tryParse should be called without checking whether the controller.text has a value. 
if (_doseController.text.isNotEmpty) {
 . .... 
 setState(()...)
}

Reply:
Done